ddl/ddl_api.go

```diff
@@ -5165,9 +5166,110 @@ func (d *ddl) AlterIndexVisibility(ctx sessionctx.Context, ident ast.Ident, inde
 	return errors.Trace(err)
 }
 
+func checkPlacementLabelConstraint(label string) (placement.LabelConstraint, error) {
+	r := placement.LabelConstraint{}
+
+	if len(label) < 4 {
+		return r, errors.Errorf("label constraint should be in format '{+|-}key=value', but got '%s'", label)
+	}
+
+	var op placement.LabelConstraintOp
+	switch label[0] {
+	case '+':
+		op = placement.In
+	case '-':
+		op = placement.NotIn
+	default:
+		return r, errors.Errorf("label constraint should be in format '{+|-}key=value', but got '%s'", label)
+	}
+
+	kv := strings.Split(label[1:], "=")
+	if len(kv) != 2 {
+		return r, errors.Errorf("label constraint should be in format '{+|-}key=value', but got '%s'", label)
+	}
+
+	key := strings.TrimSpace(kv[0])
+	if key == "" {
+		return r, errors.Errorf("label constraint should be in format '{+|-}key=value', but got '%s'", label)
+	}
+
+	val := strings.TrimSpace(kv[1])
+	if val == "" {
+		return r, errors.Errorf("label constraint should be in format '{+|-}key=value', but got '%s'", label)
+	}
+
+	r.Key = key
+	r.Op = op
+	r.Values = []string{val}
+	return r, nil
+}
+
+func checkPlacementLabelConstraints(rule *placement.Rule, labels []string) error {
+	for _, str := range labels {
+		label, err := checkPlacementLabelConstraint(strings.TrimSpace(str))
+		if err != nil {
+			return err
+		}
+		rule.LabelConstraints = append(rule.LabelConstraints, label)
+	}
+	return nil
+}
+
+func checkPlacementSpecConstraint(rules []*placement.Rule, rule *placement.Rule, cnstr string) ([]*placement.Rule, error) {
+	cnstr = strings.TrimSpace(cnstr)
+	var err error
+	if len(cnstr) > 0 && cnstr[0] == '[' {
+		constraints := []string{}
+
+		err = json.Unmarshal([]byte(cnstr), &constraints)
+		if err != nil {
+			return rules, err
+		}
+
+		err = checkPlacementLabelConstraints(rule, constraints)
+		if err != nil {
+			return rules, err
+		}
+
+		rules = append(rules, rule)
+	} else if len(cnstr) > 0 && cnstr[0] == '{' {
+		constraints := map[string]int{}
+		err = json.Unmarshal([]byte(cnstr), &constraints)
+		if err != nil {
+			return rules, err
+		}
+
+		for labels, cnt := range constraints {
+			newrule := &placement.Rule{}
+			*newrule = *rule
+			if cnt <= 0 {
+				err = errors.Errorf("count should be non-positive, but got %d", cnt)
+				break
+			}
+			// TODO: handle or remove rule.Count in later commits
+			rule.Count -= cnt
+			newrule.Count = cnt
```

Comment:
The `rule.Count` is not used in anywhere?

The `newrule.Count` is populated with `cnt`, and the `rule` is not appended to `rules`.

Reply:
Yes, it is not used for now. For a dict constrait, if `rule.Count > 0` after the iteration, I need to add the extra replicas to the default rule.
But that is not even introduced so far..It may be introduced in the next PR.

Reply:
So do we need to perfect this TODO?

Reply:
I would like to add it later... I did not even think up a good ID format for the "default rule".
This PR only serves as a transition from the old syntax to the new syntax. The syntax has changed too much. I do not want to mix all these code with the implementation of `ALTER`...